### PR TITLE
Implementing bindings for the new Image2Tensor feature

### PR DESCRIPTION
**Pull request by brianjjones**
New bindings for the optional Image2Tensor feature
of Wasmtime wasi-nn. This exposes the functions
provided by the image2tensor crate, which converts
standard images to tensors. Note: this requires
wasmtime to be compiled using --features "wasmtime-wasi-nn/i2t"
to work. If the wasmtime version you are using doesn't
have this feature, the rust bindings will still work
because they will default to using the image2tensor
crate directly. However, the AssemblyScript bindings
will not work, since image2tensor is written in rust.

### CODE REVIEW

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+     * @param target the device on which to run the graph
+     * @returns an initialized `Graph`
+     */
+    static load(builder: u8[][], encoding: GraphEncoding, target: ExecutionTarget): Graph {
+        let graphBuilder: u32[] = [];
+
+        for (let i = 0; i < builder.length; i++) {
+            graphBuilder.push(getArrayPtr(builder[i]));
+            graphBuilder.push(builder[i].length);
+        }
```
Comment (MaxGraey) [inline assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:37]:
```suggestion
        let len = builder.length;
        let graphBuilder = new Array<u32>(len * 2);

        for (let i = 0; i < len; i++) {
            let arr = builder[i];
            graphBuilder[i * 2 + 0] = getArrayPtr(arr);
            graphBuilder[i * 2 + 1] = arr.length;
        }
```

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+        let graphPointer: u32 = changetype<u32>(memory.data(4));
```
Comment (MaxGraey) [inline assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:39]:
```suggestion
        let graphPointer = memory.data(4) as u32;
```

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+     * `NotEnoughMemory` error, try upping the size of `outputBuffer`.
+     * @param index the index of the output parameter to retrieve
+     * @param outputBuffer the output buffer to use as the destination for the output bytes; once we
+     * have functions to get the output buffer size, this parameter will become unnecessary and
+     * should be removed (TODO).
+     * @returns the output tensor
+     */
+    getOutput(index: u32, outputBuffer: Array<u8>): Tensor {
+        let maxBufferLength = outputBuffer.length;
+        let bytesWritten: u32 = changetype<u32>(memory.data(4));
```
Comment (MaxGraey) [inline assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:119]:
```suggestion
        let bytesWritten = memory.data(4) as u32;
```

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+ */
+export function convert_image(path: string, width: u32, height: u32, precision: TensorType): u8[] {
+
+    let bytes = 1;
+
+    if (precision === TensorType.f32 || precision === TensorType.i32) {
+        bytes = 4;
+    } else if (precision === TensorType.f16) {
+        bytes = 2;
+    }
```
Comment (MaxGraey) [inline assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:184]:
```suggestion
    let bytes: i32;
    switch (precision) {
       case TensorType.f16:
         bytes = 2;
         break;
       case TensorType.i32:
       case TensorType.f32:
         bytes = 4;
         break;
       default:
         bytes = 1;
         break;
    }
```

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    // The 3 is because we have Red, Green, and Blue in each pixel.
+    let out_buffer = new Array<u8>(width * height * 3 * bytes).fill(0);
```
Comment (MaxGraey) [inline assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:187]:
```suggestion
    let out_buffer = new Array<u8>(width * height * 3 * bytes);
```
In AS you can skip filling array to zeros

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    let bytesWritten: i32 = changetype<u32>(memory.data(16));
```
Comment (MaxGraey) [inline assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:188]:
```suggestion
    let bytesWritten = memory.data(16) as u32;
```

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+    // Convert path to u8[]
+    for (let i = 0; i < pathTrim.length; i++) {
+        ut8[i] = pathTrim.charCodeAt(i) as u8;
+    }
+
+    let u8ptr = getArrayPtr(ut8);
+    wasi_ephemeral_nn.convert_image(u8ptr, ut8.length, width, height, precision,
+                                    getArrayPtr(out_buffer), out_buffer.length,
+                                    bytesWritten);
```
Comment (MaxGraey) [inline assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:200]:
```suggestion    
    let utf8buf = String.UTF8.encode(pathTrim);
    let utf8ptr = changetype<u32>(utf8buf);
    let utf8len = utf8buf.length;
    wasi_ephemeral_nn.convert_image(utf8ptr, utf8len, width, height, precision,
                                    getArrayPtr(out_buffer), out_buffer.length,
                                    bytesWritten);
```

File: assemblyscript/assembly/convert_image/host/object-classification.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+ * @returns an array of results
+ */
+function sortResults(output: wasi_nn.Tensor, topK: u32): Result[] {
+    const probabilities = Float32Array.wrap(output.toArrayBuffer()).slice(1);
+    const results = new Array<Result>(probabilities.length);
+    // TODO figure out why we cannot use `map` here.
+    for (let i = 0; i < probabilities.length; i++) {
+        results[i] = new Result(i, probabilities[i]);
+    }
+    results.sort((a: Result, b: Result) => a.probability > b.probability ? -1 : 1);
```
Comment (MaxGraey) [inline assemblyscript/assembly/convert_image/host/object-classification.ts:73]:
```suggestion
    results.sort((a, b) => a.probability > b.probability ? -1 : 1);
```

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts (hunk 2 above)
Comment (brianjjones) [reply assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:39]: Why is `as u32` preferred to using changetype in this situation?

Comment (MaxGraey) [reply assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:39]: `changetype` is special cast which usually uses for unsafe cast some reference to `usize` (or u32) and wise versa. `memory.data` return `usize` which is `u32` in wasm32 and `u64` on wasm64 (doesn't support yet by spec). For int <-> int cast you should use ordinal type casting via `<T>val` or `val as T`

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts (hunk 5 above)
Comment (brianjjones) [reply assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:187]: Good to know, thanks.

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts (hunk 4 above)
Comment (MaxGraey) [reply assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:184]: I changed it to `u32`. Also I recommend move this code to some util function called `bytesFromPrecision`

File: assemblyscript/assembly/convert_image/host/as-wasi-nn.ts (hunk 2 above)
Comment (brianjjones) [reply assemblyscript/assembly/convert_image/host/as-wasi-nn.ts:39]: Thanks for the info!

**Comment by brianjjones** [comment]
Closing for now. We can revisit if there's interest in the future.

**Pull request closed without merging**